crates/ruff/src/rules/flake8_pyi/rules/duplicates_in_union.rs

```diff
@@ -0,0 +1,69 @@
+use ruff_diagnostics::{AlwaysAutofixableViolation, Diagnostic, Edit};
+use ruff_macros::{derive_message_formats, violation};
+use ruff_python_ast::types::Range;
+use rustpython_parser::ast::{Expr, ExprKind, Operator};
+
+use crate::checkers::ast::Checker;
+use crate::registry::AsRule;
+
+#[violation]
+pub struct DuplicatesInUnion {
+    pub duplicate_name: String,
+}
+
+impl AlwaysAutofixableViolation for DuplicatesInUnion {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        format!("Duplicate {} in union", self.duplicate_name)
+    }
+
+    fn autofix_title(&self) -> String {
+        format!("Remove latter {} from union", self.duplicate_name)
+    }
+}
+
+///PYI016
+pub fn duplicates_in_union(checker: &mut Checker, left: &Expr, right: &Expr) {
+    // The union data structure always works like so:
+    // a | b | c | d -> (((a | b) | c) | d).
+    // This function gets called on each pair of brackets, so it's safe to only check if the
+    // right is the duplicate (since we will have already checked the others in other invocations)
+
+    // Collapse down the left side of the left expression into a vector of nodes
+    let mut left_nodes: Vec<ExprKind> = Vec::new();
```

Comment:
I was using a HashSet here, but it looks like ExprKind doesn't implement HashCode & Equals

Reply:
While large unions are not common, I'd definitely say they're possible. So an O(n^2) algorithm does seems problematic.

Reply:
We have a thing to help with this -- check out `ComparableExpr`. You can convert from `Expr` to `ComparableExpr` to get location-agnostic hash and equality comparisons.

Reply:
We use this (e.g.) for the "repeated keys in dictionary" checks (`repeated_keys`).

Reply:
Turns out I'd over-complicated things - I've reduced this part down to a search. It's still O(n^2) over the whole array though.

If it's still too slow, I think it can be sped up via:
- a guard clause if the parent is a union (so it only runs once on long unions)
- raising violations for **all** duplicates via a HashSet - if it already exists, raise a violation
- tada - O(n)
Though I feel this would be slower for short unions, due to the more complex setup. Happy to try it out if you think it's worth it?